DataFormats/HcalRecHit/interface/HBHEChannelInfo.h

```diff
@@ -0,0 +1,213 @@
+        {return (rawCharge_[ts] - pedestal_[ts])*gain_[ts];}
+    inline double tsGain(const unsigned ts) const {return gain_[ts];}
+    inline uint8_t tsAdc(const unsigned ts) const {return adc_[ts];}
+    inline float tsRiseTime(const unsigned ts) const
+        {return hasTimeInfo_ ? riseTime_[ts] : HcalSpecialTimes::UNKNOWN_T_NOTDC;}
+
+    // Signal rise time measurement for the SOI, if available
+    inline float soiRiseTime() const
+    {
+        return (hasTimeInfo_ && soi_ < nSamples_) ?
+                riseTime_[soi_] : HcalSpecialTimes::UNKNOWN_T_NOTDC;
+    }
+
+    // The TS with the "end" index is not included in the window
+    inline double chargeInWindow(const unsigned begin, const unsigned end) const
+    {
+        double sum = 0.0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            sum += (rawCharge_[i] - pedestal_[i]);
+        return sum;
+    }
+
+    inline double energyInWindow(const unsigned begin, const unsigned end) const
+    {
+        double sum = 0.0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            sum += (rawCharge_[i] - pedestal_[i])*gain_[i];
+        return sum;
+    }
+
+    // The two following methods return MAXSAMPLES if the specified
+    // window does not overlap with the samples stored
+    inline unsigned peakChargeTS(const unsigned begin, const unsigned end) const
+    {
+        unsigned iPeak = MAXSAMPLES;
+        double dmax = -DBL_MAX;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+        {
+            const double q = rawCharge_[i] - pedestal_[i];
+            if (q > dmax)
+            {
+                dmax = q;
+                iPeak = i;
+            }
+        }
+        return iPeak;
+    }
+
+    inline unsigned peakEnergyTS(const unsigned begin, const unsigned end) const
+    {
+        unsigned iPeak = MAXSAMPLES;
+        double dmax = -DBL_MAX;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+        {
+            const double e = (rawCharge_[i] - pedestal_[i])*gain_[i];
+            if (e > dmax)
+            {
+                dmax = e;
+                iPeak = i;
+            }
+        }
+        return iPeak;
+    }
+
+    // The following function can be used, for example,
+    // in a check for presence of saturated ADC values
+    inline uint8_t peakAdcValue(const unsigned begin, const unsigned end) const
+    {
+        uint8_t peak = 0;
+        const unsigned imax = end < nSamples_ ? end : nSamples_;
+        for (unsigned i=begin; i<imax; ++i)
+            if (adc_[i] > peak)
+                peak = adc_[i];
+        return peak;
+    }
+
+private:
+    HcalDetId id_;
+
+    // Charge in fC for all time slices
+    double rawCharge_[MAXSAMPLES];
```

Comment:
The interface let's the outside to use either the bare pointer
or to access the array elements individually. See the methods
whose names start with "ts". As the internals of Method 2/Method 3
use arrays, access to bare pointers will allow the new class to
be used inside Method 2/Method 3 with minimal modifications
and without an extra copy of the data.

Surely, HBHEChannelInfo wraps the C array functionality. Also,
std::array wraps the C array functionality. If you want CMSSW
developers to write wrappers for wrappers, please document
this in some guidelines. I am usually just using common sense in
these issues.

On 07/14/2016 12:30 PM, Slava Krutelyov wrote:
>
> In DataFormats/HcalRecHit/interface/HBHEChannelInfo.h <https://github.com/cms-sw/cmssw/pull/15092#discussion_r70847637>:
>
> > +    // in a check for presence of saturated ADC values
> > +    inline uint8_t peakAdcValue(const unsigned begin, const unsigned end) const
> > +    {
> > +        uint8_t peak = 0;
> > +        const unsigned imax = end < nSamples_ ? end : nSamples_;
> > +        for (unsigned i=begin; i<imax; ++i)
> > +            if (adc_[i] > peak)
> > +                peak = adc_[i];
> > +        return peak;
> > +    }
> > +
> > +private:
> > +    HcalDetId id_;
> > +
> > +    // Charge in fC for all time slices
> > +    double rawCharge_[MAXSAMPLES];
>
> Not the analyzers fault: your interface lets the outside to see only the bare pointer.
> Within the scope of this class itself if you just use the bare array, the analyzer missing the issue can be blamed.
>
> I thought of making this comment about std::array during review, but then (mistakenly?) assumed that the interfaces where this object is used force you to pass a bare pointer. In this case it didn't 
> make much sense to switch to std::array just to lose the details in the scope where HBHEChannelInfo is used.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/15092/files/cb95544ae57dc7aa6cab5481015b80942b61d6d8#r70847637>, or mute the thread 
> <https://github.com/notifications/unsubscribe/AFS4-b1ea3xyzd4AKv786QVai2ai3ZWnks5qVnJOgaJpZM4JCc97>.
>